packages/react-dom/src/events/DOMModernPluginEventSystem.js

```diff
@@ -174,6 +178,66 @@ export function dispatchEventForPluginEventSystem(
 ): void {
   let ancestorInst = targetInst;
 
+  if (rootContainer.nodeType !== DOCUMENT_NODE) {
+    let node = targetInst;
+
+    // The below logic attempts to work out if we need to change
+    // the target fiber to a different ancestor. We had similar logic
+    // in the legacy event system, except the big difference between
+    // systems is that the modern event system now has an event listener
+    // attached to each React Root and React Portal Root. Together,
+    // the DOM nodes representing these roots are the "rootContainer".
+    // To figure out which ancestor instance we should use, we traverse
+    // up the fiber tree from the target instance and attempt to find
+    // root boundaries that match that of our current "rootContainer".
+    // If we find that "rootContainer", we find the parent fiber
+    // sub-tree for that root and make that our ancestor instance.
+    while (true) {
+      if (node === null) {
+        return;
+      } else if (node.tag === HostRoot || node.tag === HostPortal) {
+        const container = node.stateNode.containerInfo;
+        if (
+          container === rootContainer ||
+          (container.nodeType === COMMENT_NODE &&
+            container.parentNode === rootContainer)
+        ) {
+          break;
+        }
+        if (node.tag === HostPortal) {
+          // The target is a portal, but it's not the rootContainer we're looking for.
+          // Normally portals handle their own events all the way down to the root.
+          // So we should be able to stop now. However, we don't know if this portal
+          // was part of *our* root.
+          let grandNode = node.return;
+          while (grandNode !== null) {
+            if (grandNode.tag === HostRoot || grandNode.tag === HostPortal) {
+              const grandContainer = grandNode.stateNode.containerInfo;
+              if (
+                grandContainer === rootContainer ||
+                (grandContainer.nodeType === COMMENT_NODE &&
+                  grandContainer.parentNode === rootContainer)
```

Comment:
This is a repeat of the same code about 17 lines up. It could be a separate function.